tests/tests/helpers.rs

```diff
@@ -249,3 +253,89 @@ pub async fn pipe_from_stream_and_drop<T: Serialize>(
 		}
 	}
 }
+
+#[allow(dead_code)]
+pub async fn socks_server_no_auth() -> SocketAddr {
+	let mut config = server::Config::default();
+	config.set_dns_resolve(false);
+	let config = std::sync::Arc::new(config);
+
+	let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
+	let proxy_addr = listener.local_addr().unwrap();
+
+	spawn_socks_server(listener, config).await;
+
+	proxy_addr
+}
+
+#[allow(dead_code)]
+pub async fn spawn_socks_server(listener: tokio::net::TcpListener, config: std::sync::Arc<server::Config>) {
+	let addr = listener.local_addr().unwrap();
+	tokio::spawn(async move {
+		loop {
+			let (stream, _) = listener.accept().await.unwrap();
+			let mut socks5_socket = server::Socks5Socket::new(stream, config.clone());
+			socks5_socket.set_reply_ip(addr.ip());
+
+			socks5_socket.upgrade_to_socks5().await.unwrap();
+		}
+	});
+}
+
+#[pin_project(project = DataStreamProj)]
+#[allow(dead_code)]
+pub enum DataStream<T: tokio::io::AsyncRead + tokio::io::AsyncWrite + std::marker::Unpin> {
```

Comment:
I don't understand why this type is needed, can't you just use `fast_socks5::client::Socks5Stream` directly instead?

It already implements AsyncRead and AsyncWrite?

Reply:
@niklasad1 It does implement the `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`, but currently the [`TransportSenderT` and `TransportReceiverT`](https://github.com/paritytech/jsonrpsee/pull/1168/files#diff-a8985d79c808da60a2c0bd792d8d0e303791e5c1618215e764f582d8c75de292R228-R271) requires the `futures::io::AsyncRead` and `futures::io::AsyncRead` implementation.

I made this way not to change the current implementation of [`EitherStream`](https://github.com/paritytech/jsonrpsee/blob/ac0caafd4edae89f4d0aab15ccfedb93cb927dfc/client/transport/src/ws/stream.rs#L41-L154), which implements the `futures::io` traits. 

I can change the `EitherStream` to use the `tokio::io` traits instead, do you see any problem ?

> I'm actually biased because requiring the tokio traits as default would make it easier for me on other projects 😅

Reply:
This is just for tests so I don't mind having this extra code.

> I can change the EitherStream to use the tokio::io traits instead, do you see any problem ?

Let's keep it out this PR but `tokio::io` makes sense to me since this crate it's tightly-coupled to tokio nowdays.
This is probably old code when this was supposed to support both tokio and async-std.

//cc @lexnv @jsdw

Reply:
Agreed that moving to `tokio::io` traits probably makes sense and might simplify some stuff given that we are "all in" on tokio in this lib nowadays

Reply:
Also I wonder why this is an enum and not struct?

Reply:
yeah, would be cleaner with a struct to avoid matching on it :P

Reply:
Great, thanks for the suggestion and @niklasad1 for addressing the changes